### ISSUE TITLE
Setting server.context-path breaks login without trailing slash

### ISSUE THREAD

**Issue by robertdale** [spring-projects/spring-boot#4937]
This may be related to #4853 .

With server.context-path=/user set in application.properties, one could have a secured "/" and be able to login.  v1.3.1 breaks this behavior by preventing log in and automatically redirecting back to /.

If paths are secure with (secured / but open css resources):
```
http.authorizeRequests()
                .antMatchers("/css/**").permitAll()
                 .anyRequest().authenticated()
                 .and().formLogin()
...
```

v1.3.0 this would work:
enter:  http://localhost:8080/user
redirect:  http://localhost:8080/user/login
redirect:  http://localhost:8080/user/

In v1.3.1, you are redirected back to the login page:
enter: http://localhost:8080/user
redirect:  http://localhost:8080/user/login
redirect:  http://localhost:8080/user/login

It appears that the http session is cleared.  A new session cookie is created, security context cleared, etc.

Now, in v1.3.1, if you access the URL with the trailing /, then it works as expected.
enter: http://localhost:8080/user/
redirect:  http://localhost:8080/user/login
redirect:  http://localhost:8080/user/

See my repo for working examples.  Use the "complete" module.

https://github.com/robertdale/gs-securing-web/tree/1.3.1-broken

https://github.com/robertdale/gs-securing-web/tree/1.3.0-working

**Comment by robertdale** [comment]
May be related to #4715  with upgrade of tomcat from 8.0.28 to 8.0.30 and additional code:
```
		try {
			context.setUseRelativeRedirects(false);
		}
		catch (NoSuchMethodError ex) {
			// Tomcat is < 8.0.30. Continue
		}
```

And just to clarify the redirect issue:

v1.3.0 with tomcat v8.0.28
begin:    http://localhost:8080/foo
redirect: http://localhost:8080/foo/
redirect: http://localhost:8080/foo/login  (do login)
redirect: http://localhost:8080/foo/

v1.3.1 with tomcat v8.0.30
begin:    http://localhost:8080/foo
redirect: http://localhost:8080/foo/login  (do login)
redirect: http://localhost:8080/foo
redirect: http://localhost:8080/foo/login

**Comment by wilkinsona** [comment]
That shouldn't make any difference. Tomcat's default changed in 8.0.30 so the extra code changes it back so that 8.0.30 works in the same way as earlier versions.

**Comment by robertdale** [comment]
Just documenting some things for later reference on the issue...
https://bz.apache.org/bugzilla/show_bug.cgi?id=56917
https://bz.apache.org/bugzilla/show_bug.cgi?id=58660
https://bz.apache.org/bugzilla/show_bug.cgi?id=58765

Konstantin Kolinko:
>  You can opt-in for the old behaviour by setting mapperContextRootRedirectEnabled="true"
>  either on your own web application or globally in conf/context.xml file.

Mark Thomas:
> Fixed in 8.0.x (for 8.0.31 onwards)

**Comment by robertdale** [comment]
Workaround is:

```
@Configuration
public class TomcatConfig {

  @Bean
  public EmbeddedServletContainerFactory servletContainer() {
    TomcatEmbeddedServletContainerFactory factory = new TomcatEmbeddedServletContainerFactory();
    factory.setTomcatContextCustomizers(Arrays.asList(new CustomCustomizer()));
    return factory;
  }

  public static class CustomCustomizer implements TomcatContextCustomizer {
    public void customize(Context context) {
      context.setMapperContextRootRedirectEnabled(true);
    }
  }
}
```

**Comment by wilkinsona** [comment]
@robertdale Many thanks for tracking this down. It sounds like we either need to add your workaround or, if the dates work out and it's released in time, upgrade to 8.0.31 in our next maintenance release.
